analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Collections.cs

```diff
@@ -0,0 +1,79 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors;
+
+internal sealed partial class Binary
+{
+    private ProgramState LearnBranchingCollectionConstraint(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch)
+    {
+        IOperation otherOperand;
+        bool shouldFlipOperator;
+        if (InstanceOfCountPropertyReference(binary.LeftOperand) is { } collection)
+        {
+            otherOperand = binary.RightOperand;
+            shouldFlipOperator = true;
+        }
+        else
+        {
+            otherOperand = binary.LeftOperand;
+            shouldFlipOperator = false;
+            collection = InstanceOfCountPropertyReference(binary.RightOperand);
+        }
+
+        return collection is not null
+            && state.Constraint<NumberConstraint>(otherOperand) is { } number
+            && BranchingCollectionConstraint(binary.OperatorKind, falseBranch, shouldFlipOperator, number) is { } constraint
+                ? state.SetSymbolConstraint(collection, constraint)
+                : state;
+
+        ISymbol InstanceOfCountPropertyReference(IOperation operation) =>
+            operation.AsPropertyReference() is { Instance: { } instance, Property.Name: nameof(Array.Length) or nameof(List<int>.Count) }
+            && instance.TrackedSymbol(state) is { } symbol
+                ? symbol
+                : null;
+    }
+
+    private static SymbolicConstraint BranchingCollectionConstraint(BinaryOperatorKind operatorKind, bool falseBranch, bool countIsLeftOperand, NumberConstraint number)
```

Comment:
```suggestion
    private static SymbolicConstraint SetCollectionConstraint(BinaryOperatorKind operatorKind, bool falseBranch, bool shouldFlipOperator, NumberConstraint number)
```

or

```
    private static SymbolicConstraint  ComputeCollectionConstraint
```
or something else/